server.go

```diff
@@ -155,11 +158,25 @@ func (server *mongoServer) AcquireSocket(poolLimit int, minPoolSize int, timeout
 					socket.Close()
 					return nil, abended, errServerClosed
 				}
+				// Replace a nil placeholder with the new socket,
+				// it does not matter which nil-placeholder we replace.
+				//
+				// The reason we do not publish the socket in liveSockets until after
+				// connection is completed, is that an unconnected socket is not safe
+				// for concurrent use. Concurrent mongoServer.Close() calls may race with us
+				// and otherwise call Close() on an unlocked not-yet-connected socket.
+				for i, s := range server.liveSockets {
```

Comment:
I feel slightly uncomfortable with the fact that this is O(n) during connection acquisition (i.e. O(n^2) to open n new connections).

However, the constant cost is low and the connection caps we use are relatively small. I'd rather us pay this cost than go with a more complicated fix with a higher risk of introducing a bug.

Reply:
I thought about that, but I think we'll have much bigger problems if the connection count becomes so large that this is unreasonably slow.  I'm pretty sure it would need to be in the millions before this takes anywhere near a millisecond (searching 1mil pointers took ~500μs on my machine, example test code below).

```golang
package main

import (
	"math/rand"
	"testing"
	"time"
)

func init() {
	rand.Seed(time.Now().UnixNano())
}

// Fills space to make the test a little more realistic
type Deadweight struct {
	A uint64
	_ uint64
	_ uint64
	_ uint64
	_ uint64
	_ uint64
	_ uint64
	_ uint64
}

func FindNil(a []*Deadweight) int {
	for i, p := range a {
		if p == nil {
			return i
		}
	}
	return -1
}

func BenchmarkPointer(b *testing.B) {
	const N = 1000000
	a := make([]*Deadweight, N)
	for i := range a {
		a[i] = &Deadweight{A: uint64(i)}
	}
	n := rand.Intn(1000) // place near end
	a[len(a)-n-1] = nil
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		FindNil(a)
	}
}

```

Reply:
Agreed. Sorry, I didn't meant to make you do a test :)

Reply:
No worries, I was actually curious about it how long it would take.